server/streamable_http.go

```diff
@@ -0,0 +1,812 @@
+	lastEventID := r.Header.Get("Last-Event-Id")
+	httpSession, ok := session.(*streamableHTTPSession)
+	if lastEventID != "" && ok && httpSession.eventStore != nil {
+		// Replay events that occurred after the last event ID
+		err := httpSession.eventStore.ReplayEventsAfter(lastEventID, func(eventID string, message mcp.JSONRPCMessage) error {
+			data, err := json.Marshal(message)
+			if err != nil {
+				return err
+			}
+
+			// Write the event directly to the response writer
+			fmt.Fprintf(w, "id: %s\ndata: %s\n\n", eventID, data)
+			w.(http.Flusher).Flush()
+			return nil
+		})
+
+		if err != nil {
+			// Log the error but continue
+			fmt.Printf("Error replaying events: %v\n", err)
+		}
+	}
+
+	// Send the initial response if there is one
+	if initialResponse != nil {
+		data, err := json.Marshal(initialResponse)
+		if err != nil {
+			http.Error(w, "Failed to marshal response", http.StatusInternalServerError)
+			return
+		}
+
+		// Store the event if we have an event store
+		var eventID string
+		if httpSession != nil && httpSession.eventStore != nil {
+			var storeErr error
+			eventID, storeErr = httpSession.eventStore.StoreEvent(streamID, initialResponse)
+			if storeErr != nil {
+				// Log the error but continue
+				fmt.Printf("Error storing event: %v\n", storeErr)
+			}
+		}
+
+		// Write the event directly to the response writer
+		if eventID != "" {
+			fmt.Fprintf(w, "id: %s\ndata: %s\n\n", eventID, data)
+		} else {
+			fmt.Fprintf(w, "data: %s\n\n", data)
+		}
+		w.(http.Flusher).Flush()
+	}
+
+	// Create a channel to pass notifications from the goroutine to the main handler
+	notificationCh := make(chan struct {
+		eventID string
+		data    []byte
+	}, 100) // Buffer size to prevent blocking
+	notifDone := make(chan struct{})
+	defer close(notifDone)
+
+	// Start a goroutine to listen for notifications and send them to the notification channel
+	go func() {
+		for {
+			select {
+			case notification, ok := <-httpSession.notificationChannel:
+				if !ok {
+					return
+				}
+
+				data, err := json.Marshal(notification)
+				if err != nil {
+					continue
+				}
+
+				// Store the event if we have an event store
+				var eventID string
+				if httpSession != nil && httpSession.eventStore != nil {
+					var storeErr error
+					eventID, storeErr = httpSession.eventStore.StoreEvent(streamID, notification)
+					if storeErr != nil {
+						// Log the error but continue
+						fmt.Printf("Error storing event: %v\n", storeErr)
+					}
+				}
+
+				// Send the notification to the main handler goroutine via channel
+				select {
+				case notificationCh <- struct {
+					eventID string
+					data    []byte
+				}{eventID: eventID, data: data}:
+				case <-notifDone:
+					return
+				}
+			case <-notifDone:
+				return
+			}
+		}
+	}()
+
+	// Create a context with cancellation
+	ctx, cancel := context.WithCancel(r.Context())
+	defer cancel()
+
+	// Process notifications in the main handler goroutine
+	for {
```

Comment:
Maybe I mis-understand the protocol specification, is the sse steam should be ended when the response is sent?

<img width="601" alt="image" src="https://github.com/user-attachments/assets/1bf0c84e-0d6f-4b1e-a0f5-0526cefaf9af" />

Reply:
Thanks for you feedback. Please see commit for this: https://github.com/mark3labs/mcp-go/pull/228/commits/e016e85080be9359ac8f8097dc46bd04370be3e1